src/neo/Ledger/Blockchain.cs

```diff
@@ -415,28 +415,36 @@ private void Persist(Block block)
                     Context.System.EventStream.Publish(application_executed);
                     all_application_executed.Add(application_executed);
                 }
-                foreach (IPersistencePlugin plugin in Plugin.PersistencePlugins)
-                    plugin.OnPersist(system, block, snapshot, all_application_executed);
-                snapshot.Commit();
-                List<Exception> commitExceptions = null;
                 foreach (IPersistencePlugin plugin in Plugin.PersistencePlugins)
                 {
-                    try
+                    if (plugin.CurrentHeight < block.Index)
                     {
-                        plugin.OnCommit(system, block, snapshot);
+                        plugin.OnPersist(system, block, snapshot, all_application_executed);
                     }
-                    catch (Exception ex)
+                }
+                List<Exception> commitExceptions = null;
+                foreach (IPersistencePlugin plugin in Plugin.PersistencePlugins)
+                {
+                    if (plugin.CurrentHeight < block.Index)
                     {
-                        if (plugin.ShouldThrowExceptionFromCommit(ex))
+                        try
                         {
-                            if (commitExceptions == null)
-                                commitExceptions = new List<Exception>();
+                            plugin.OnCommit(system, block, snapshot);
+                        }
+                        catch (Exception ex)
+                        {
+                            if (plugin.ShouldThrowExceptionFromCommit(ex))
+                            {
+                                if (commitExceptions == null)
+                                    commitExceptions = new List<Exception>();
 
-                            commitExceptions.Add(ex);
+                                commitExceptions.Add(ex);
+                            }
                         }
                     }
                 }
                 if (commitExceptions != null) throw new AggregateException(commitExceptions);
+                snapshot.Commit();
```

Comment:
If snapshot is the last to commit then https://github.com/neo-project/neo/issues/2570#issuecomment-900178199 is probably not relevant though, plugins won't ever have N-1 data, rather they could have some N+1, but they won't process it again because of `plugin.CurrentHeight`. So it should be fine.